File: 051/1.js
// Interface Payable
class Payable {
    processPayment() {
        throw new Error("Method 'processPayment()' must be implemented.");
    }
}

// Employee class
class Employee extends Payable {
    constructor(name, id, position, salary) {
        super();
        this._name = name;
        this._id = id;
        this._position = position;
        this._salary = salary;
    }

    // Getter and Setter for name
    getName() {
        return this._name;
    }

    setName(name) {
        this._name = name;
    }

    // Getter and Setter for id
    getId() {
        return this._id;
    }

    setId(id) {
        this._id = id;
    }

    // Getter and Setter for position
    getPosition() {
        return this._position;
    }

    setPosition(position) {
        this._position = position;
    }

    // Getter and Setter for salary
    getSalary() {
        return this._salary;
    }

    setSalary(salary) {
        this._salary = salary;
    }

    displayDetails() {
        console.log(`ชื่อ: ${this._name}, รหัส: ${this._id}, ตำแหน่ง: ${this._position}, ค่าจ้าง: ${this._salary}`);
    }

    calculateBonus(percentage) {
        const bonus = this._salary * (percentage / 100);
        console.log(`โบนัส: ${bonus}`);
        return bonus;
    }

    processPayment() {
        console.log(`ประมวลผลการชำระเงินสำหรับ ${this._name}`);
    }
}

// Manager class
class Manager extends Employee {
    constructor(name, id, position, salary, specialBonus) {
        super(name, id, position, salary);
        this._specialBonus = specialBonus;
    }

    calculateBonus(percentage) {
        const baseBonus = super.calculateBonus(percentage);
        const totalBonus = baseBonus + this._specialBonus;
        console.log(`โบนัสรวม (รวมโบนัสพิเศษ): ${totalBonus}`);
        return totalBonus;
    }
}

// Demonstration of Employee Management System
const emp = new Employee("ณทชัย", 101, "นักพัฒนา", 50000);
const manager = new Manager("ชัยทณ", 102, "ผู้จัดการ", 80000, 5000);

emp.displayDetails();
emp.calculateBonus(10);
emp.processPayment();

console.log("\n");

manager.displayDetails();
manager.calculateBonus(10);
manager.processPayment();

File: 051/1.py

from abc import ABC, abstractmethod

class Payable(ABC):
    @abstractmethod
    def processPayment(self):
        pass

class Employee(Payable):
    def __init__(self, name, id, position, salary):
        self.__name = name
        self.__id = id
        self.__position = position
        self.__salary = salary


    def getName(self):
        return self.__name

    def setName(self, name):
        self.__name = name

 
    def getId(self):
        return self.__id

    def setId(self, id):
        self.__id = id


    def getPosition(self):
        return self.__position

    def setPosition(self, position):
        self.__position = position

 
    def getSalary(self):
        return self.__salary

    def setSalary(self, salary):
        self.__salary = salary

    def displayDetails(self):
        print(f"ชื่อ: {self.__name}, รหัส: {self.__id}, ตำเเหน่ง: {self.__position}, ค่่าจ้าง: {self.__salary}")

    def calculateBonus(self, percentage):
        bonus = self.__salary * (percentage / 100)
        print(f"โบนัส: {bonus}")
        return bonus

    def processPayment(self):
        print(f"ประมวลผลการชำระเงินสำหรับ {self.__name}")

class Manager(Employee):
    def __init__(self, name, id, position, salary, specialBonus):
        super().__init__(name, id, position, salary)
        self.__specialBonus = specialBonus

    def calculateBonus(self, percentage):
        baseBonus = super().calculateBonus(percentage)
        totalBonus = baseBonus + self.__specialBonus
        print(f"โบนัสรวม (รวมโบนัสพิเศษ): {totalBonus}")
        return totalBonus


emp = Employee("ณทชัย", 101, "นักพัฒนา", 50000)
manager = Manager("ชัยทณ", 102, "ผู้จัดการ", 80000, 5000)

emp.displayDetails()
emp.calculateBonus(10)
emp.processPayment()
manager.displayDetails()
manager.calculateBonus(10)
manager.processPayment()